File: integrations/agent-fabric/src/lease.rs
//! Exclusive task leases with TTL expiry and affine handles.
//!
//! The lease table is where "no duplicate assignment" stops being a promise and becomes a
//! structure: at most one live lease per task exists because [`LeaseTable::grant`] refuses while
//! one is held, and the returned [`LeaseHandle`] is deliberately **not `Clone`** — releasing or
//! renewing consumes it, so double-release and grant-then-forget are compile errors, not runtime
//! audits. Expiry is how the fabric detects crashed or silently-dead workers: the lease simply
//! outlives its TTL and the task returns to the retry path.

use crate::ids::{AgentId, LeaseEpoch, TaskId};
use std::collections::BTreeMap;
use std::fmt;

#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub enum LeaseError {
    /// Another agent holds a live lease on the task.
    HeldByOther {
        holder: AgentId,
    },
    /// The presented handle's epoch no longer matches — the lease expired and was re-granted.
    EpochMismatch,
    UnknownTask,
}

impl fmt::Display for LeaseError {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            LeaseError::HeldByOther { holder } => write!(f, "task already leased to {holder}"),
            LeaseError::EpochMismatch => write!(f, "lease epoch no longer current"),
            LeaseError::UnknownTask => write!(f, "no lease recorded for this task"),
        }
    }
}

impl std::error::Error for LeaseError {}

/// Ownership token for one live lease. Not `Clone`, not `Copy`: holding one *is* holding the
/// exclusive right, and rights that can be duplicated are not exclusive.
#[derive(Debug, PartialEq, Eq)]
pub struct LeaseHandle {
    pub(crate) task: TaskId,
    pub(crate) agent: AgentId,
    pub(crate) epoch: LeaseEpoch,
}

impl LeaseHandle {
    pub fn task(&self) -> TaskId {
        self.task
    }

    pub fn agent(&self) -> AgentId {
        self.agent
    }

    pub fn epoch(&self) -> LeaseEpoch {
        self.epoch
    }
}

#[derive(Debug)]
struct Active {
    agent: AgentId,
    epoch: LeaseEpoch,
    expires_at: u64,
}

/// A released or expired lease, reported back so the scheduler can retry with full information.
#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub struct ExpiredLease {
    pub task: TaskId,
    pub agent: AgentId,
    pub ended_at_tick: u64,
}

#[derive(Debug, Default)]
pub struct LeaseTable {
    by_task: BTreeMap<TaskId, Active>,
    next_epoch: u64,
}

impl LeaseTable {
    pub fn new() -> Self {
        Self::default()
    }

    /// Grants an exclusive lease or fails with the current holder. Monotone epochs mean a stale
    /// handle from an expired generation can never be mistaken for the current one.
    pub fn grant(
        &mut self,
        task: TaskId,
        agent: AgentId,
        now: u64,
        ttl_ticks: u64,
    ) -> Result<LeaseHandle, LeaseError> {
        assert!(ttl_ticks > 0, "a zero-ttl lease would expire before use");
        if let Some(active) = self.by_task.get(&task) {
            return Err(LeaseError::HeldByOther {
                holder: active.agent,
            });
        }
        self.next_epoch += 1;
        let epoch = LeaseEpoch::new(self.next_epoch);
        self.by_task.insert(
            task,
            Active {
                agent,
                epoch,
                expires_at: now + ttl_ticks,
            },
        );
        Ok(LeaseHandle { task, agent, epoch })
    }

    /// Extends a live lease; consumes and re-mints the handle so exactly one token remains.
    pub fn renew(
        &mut self,
        h: LeaseHandle,
        now: u64,
        ttl_ticks: u64,
    ) -> Result<LeaseHandle, LeaseError> {
        assert!(ttl_ticks > 0, "a zero-ttl lease would expire before use");
        let Some(active) = self.by_task.get_mut(&h.task) else {
            return Err(LeaseError::UnknownTask);
        };
        if active.epoch != h.epoch {
            return Err(LeaseError::EpochMismatch);
        }
        active.expires_at = now + ttl_ticks;
        Ok(LeaseHandle {
            task: h.task,
            agent: h.agent,
            epoch: h.epoch,
        })
    }

    /// Ends the lease. The handle is consumed; there is no way to release twice.
    pub fn release(&mut self, h: LeaseHandle) -> Result<(), LeaseError> {
        self.release_by(h.task(), h.epoch())
    }

    /// Epoch-keyed release for callers that tracked the epoch without retaining the affine
    /// handle (the settlement path). A stale epoch is rejected, so a late completion cannot
    /// free a lease it no longer owns.
    pub fn release_by(&mut self, task: TaskId, epoch: LeaseEpoch) -> Result<(), LeaseError> {
        match self.by_task.get(&task) {
            None => Err(LeaseError::UnknownTask),
            Some(a) if a.epoch != epoch => Err(LeaseError::EpochMismatch),
            Some(_) => {
                self.by_task.remove(&task);
                Ok(())
            }
        }
    }

    /// Drops every lease whose TTL has passed by `now` and reports them for retry handling.
    pub fn expire_before(&mut self, now: u64) -> Vec<ExpiredLease> {
        let due: Vec<TaskId> = self
            .by_task
            .iter()
            .filter(|(_, a)| a.expires_at <= now)
            .map(|(t, _)| *t)
            .collect();
        due.into_iter()
            .filter_map(|t| {
                let a = self.by_task.remove(&t)?;
                Some(ExpiredLease {
                    task: t,
                    agent: a.agent,
                    ended_at_tick: a.expires_at,
                })
            })
            .collect()
    }

    pub fn holder(&self, task: TaskId) -> Option<AgentId> {
        self.by_task.get(&task).map(|a| a.agent)
    }

    pub fn live(&self) -> usize {
        self.by_task.len()
    }

    pub fn expiry_of(&self, task: TaskId) -> Option<u64> {
        self.by_task.get(&task).map(|a| a.expires_at)
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn ids(t: u64, a: u64) -> (TaskId, AgentId) {
        (
            TaskId::from_raw(t).expect("nonzero"),
            AgentId::from_raw(a).expect("nonzero"),
        )
    }

    #[test]
    fn a_second_grant_while_held_names_the_holder_rather_than_succeeding() {
        let mut t = LeaseTable::new();
        let (task, x, y) = (ids(1, 1).0, ids(1, 1).1, ids(1, 2).1);
        let h = t.grant(task, x, 0, 10).expect("first grant");
        assert_eq!(
            t.grant(task, y, 1, 10),
            Err(LeaseError::HeldByOther { holder: x })
        );
        t.release(h).expect("release");
        assert!(t.grant(task, y, 2, 10).is_ok(), "free after release");
    }

    #[test]
    fn a_consumed_handle_cannot_release_twice_because_ownership_moves() {
        let mut t = LeaseTable::new();
        let (task, x) = ids(5, 5);
        let h = t.grant(task, x, 0, 10).expect("grant");
        t.release(h).expect("first release");
        // `h` has been moved into release(); a second release(h) would not compile. The
        // observable consequence is asserted instead: nothing is left to release.
        assert_eq!(t.live(), 0);
        assert_eq!(t.holder(task), None);
    }

    #[test]
    fn renewal_consumes_the_old_handle_and_keeps_exactly_one_live_token() {
        let mut t = LeaseTable::new();
        let (task, x) = ids(7, 7);
        let h1 = t.grant(task, x, 0, 10).expect("grant");
        let h2 = t.renew(h1, 5, 20).expect("renew");
        assert_eq!(t.expiry_of(task), Some(25));
        t.release(h2).expect("release renewed");
        assert_eq!(t.live(), 0);
    }

    #[test]
    fn an_expired_generation_handle_is_rejected_not_treated_as_current() {
        let mut t = LeaseTable::new();
        let (task, x) = ids(9, 9);
        let h1 = t.grant(task, x, 0, 1).expect("grant");
        let old_epoch = h1.epoch();
        let expired = t.expire_before(5);
        assert_eq!(expired.len(), 1);
        assert_eq!(t.renew(h1, 6, 10), Err(LeaseError::UnknownTask));
        // Re-grant after expiry mints a new epoch; the old handle stays invalid forever.
        let h2 = t.grant(task, x, 6, 10).expect("re-grant");
        assert_ne!(old_epoch, h2.epoch());
    }

    #[test]
    fn expiry_reports_only_what_actually_lapsed_with_its_end_tick() {
        let mut t = LeaseTable::new();
        let (t1, a1) = ids(11, 11);
        let (t2, a2) = ids(12, 12);
        t.grant(t1, a1, 0, 10).expect("g1");
        t.grant(t2, a2, 0, 100).expect("g2");
        let out = t.expire_before(10);
        assert_eq!(
            out,
            vec![ExpiredLease {
                task: t1,
                agent: a1,
                ended_at_tick: 10
            }]
        );
        assert_eq!(t.live(), 1);
    }
}

File: ports/go-control-plane/lease.go
package controlplane

import (
	"errors"
	"fmt"
	"sort"
)

// ErrZeroTTL reports a lease grant or renewal whose TTL is zero: such a lease
// would expire before use, so it is refused instead of silently created.
var ErrZeroTTL = errors.New("lease ttl must be positive")

// ErrUnknownTask reports that no lease is recorded for the task at all.
var ErrUnknownTask = errors.New("no lease recorded for this task")

// HeldByOtherError reports a refused grant and names the current holder, so
// the caller learns who owns the task rather than merely that it failed.
type HeldByOtherError struct {
	Task   TaskID
	Holder AgentID
}

func (e *HeldByOtherError) Error() string {
	return fmt.Sprintf("task %s already leased to %s", e.Task, e.Holder)
}

// EpochMismatchError reports that the presented epoch is not current — the
// lease expired and was re-granted, so a stale generation can neither renew
// nor release what it once held.
type EpochMismatchError struct {
	Task      TaskID
	Presented LeaseEpoch
	Current   LeaseEpoch
}

func (e *EpochMismatchError) Error() string {
	return fmt.Sprintf("lease epoch %s no longer current for %s (current %s)",
		e.Presented, e.Task, e.Current)
}

// LeaseHandle carries the identity of one live lease. Go has no move
// semantics, so unlike the Rust handle this value is copyable; exclusivity is
// therefore enforced by epoch checks on every mutating call rather than by
// consuming the token. The consequence is stated in the docs: a caller that
// retains a handle after release gets an EpochMismatchError, not a compile
// error.
type LeaseHandle struct {
	task  TaskID
	agent AgentID
	epoch LeaseEpoch
}

// Task returns the leased task.
func (h LeaseHandle) Task() TaskID { return h.task }

// Agent returns the holder the lease was granted to.
func (h LeaseHandle) Agent() AgentID { return h.agent }

// Epoch returns the generation minted for this lease.
func (h LeaseHandle) Epoch() LeaseEpoch { return h.epoch }

type activeLease struct {
	holder    AgentID
	epoch     LeaseEpoch
	expiresAt uint64
}

// ExpiredLease reports one lapsed lease for retry handling.
type ExpiredLease struct {
	Task        TaskID
	Agent       AgentID
	EndedAtTick uint64
}

// LeaseTable grants exclusive task leases with TTL expiry. At most one live
// lease per task exists because Grant refuses while one is held, and epochs
// advance monotonically across the whole table, so a stale handle from an
// expired generation can never be mistaken for the current one.
//
// Expiry is how the control plane detects crashed or silently-dead workers:
// the lease simply outlives its TTL and the task returns to the retry path.
type LeaseTable struct {
	byTask    map[TaskID]activeLease
	nextEpoch uint64
}

// NewLeaseTable constructs an empty table.
func NewLeaseTable() *LeaseTable {
	return &LeaseTable{byTask: make(map[TaskID]activeLease)}
}

// Grant grants an exclusive lease or fails naming the current holder.
func (t *LeaseTable) Grant(task TaskID, agent AgentID, now, ttlTicks uint64) (LeaseHandle, error) {
	if ttlTicks == 0 {
		return LeaseHandle{}, ErrZeroTTL
	}
	if active, held := t.byTask[task]; held {
		return LeaseHandle{}, &HeldByOtherError{Task: task, Holder: active.holder}
	}
	t.nextEpoch++
	epoch := newLeaseEpoch(t.nextEpoch)
	t.byTask[task] = activeLease{holder: agent, epoch: epoch, expiresAt: now + ttlTicks}
	return LeaseHandle{task: task, agent: agent, epoch: epoch}, nil
}

// Renew extends a live lease. The fresh handle replaces the presented one;
// because Go copies values, both remain usable in the caller's scope, but only
// the epoch the table currently records can ever renew or release again.
func (t *LeaseTable) Renew(h LeaseHandle, now, ttlTicks uint64) (LeaseHandle, error) {
	if ttlTicks == 0 {
		return LeaseHandle{}, ErrZeroTTL
	}
	active, ok := t.byTask[h.task]
	if !ok {
		return LeaseHandle{}, ErrUnknownTask
	}
	if active.epoch != h.epoch {
		return LeaseHandle{}, &EpochMismatchError{Task: h.task, Presented: h.epoch, Current: active.epoch}
	}
	active.expiresAt = now + ttlTicks
	t.byTask[h.task] = active
	return h, nil
}

// Release ends the lease identified by the handle's task and epoch.
func (t *LeaseTable) Release(h LeaseHandle) error {
	return t.ReleaseBy(h.task, h.epoch)
}

// ReleaseBy is the epoch-keyed release for callers that tracked the epoch
// without retaining a handle (the settlement path). A stale epoch is rejected,
// so a late completion cannot free a lease it no longer owns.
func (t *LeaseTable) ReleaseBy(task TaskID, epoch LeaseEpoch) error {
	active, ok := t.byTask[task]
	if !ok {
		return ErrUnknownTask
	}
	if active.epoch != epoch {
		return &EpochMismatchError{Task: task, Presented: epoch, Current: active.epoch}
	}
	delete(t.byTask, task)
	return nil
}

// ExpireBefore drops every lease whose TTL has passed by now and reports them
// in ascending task order, matching the ordered maps of the Rust reference so
// two runs observe identical expiry streams.
func (t *LeaseTable) ExpireBefore(now uint64) []ExpiredLease {
	var due []ExpiredLease
	for task, active := range t.byTask {
		if active.expiresAt <= now {
			due = append(due, ExpiredLease{Task: task, Agent: active.holder, EndedAtTick: active.expiresAt})
		}
	}
	sort.Slice(due, func(i, j int) bool { return due[i].Task.raw < due[j].Task.raw })
	for _, d := range due {
		delete(t.byTask, d.Task)
	}
	return due
}

// Holder reports the current holder of a live lease, if any.
func (t *LeaseTable) Holder(task TaskID) (AgentID, bool) {
	active, ok := t.byTask[task]
	return active.holder, ok
}

// Live counts live leases.
func (t *LeaseTable) Live() int { return len(t.byTask) }

// ExpiryOf reports when the task's lease lapses.
func (t *LeaseTable) ExpiryOf(task TaskID) (uint64, bool) {
	active, ok := t.byTask[task]
	return active.expiresAt, ok
}
